storage/innobase/trx/trx0roll.cc

```diff
@@ -796,7 +804,23 @@ void trx_rollback_all_recovered(void*)
 	if (trx_sys.rw_trx_hash.size()) {
 		ib::info() << "Starting in background the rollback of"
 			" recovered transactions";
-		trx_rollback_recovered(true);
+
+		DBUG_EXECUTE_IF("block_rollback_thread",
+			std::this_thread::sleep_for(std::chrono::hours(100)););
+
+		bool finished = false;
+		do {
+			/* trx_rollback_all_recovered is called after finished is tested,
+			it guarantees that all active transactions are handled. */
+			finished = ddl_recovery_done || srv_shutdown_state != SRV_SHUTDOWN_NONE;
+
+			trx_rollback_recovered(true);
+
+			/* Sleep 10ms to wait for MySQL recover to revert some prepared
+			transactions to active state. */
+			std::this_thread::sleep_for(std::chrono::milliseconds(10));
+		} while (!finished);
```

Comment:
This looks prone to race conditions. I think that it would be better to initiate another rollback in or after `innodb_ddl_recovery_done()`.

Reply:
Hi Marko,
Thanks for the quick and thorough review. 

Starting a new rollback thread seems simpler. But there are two problems. 

First, at the time, the original rollback thread may not exit yet. Thus there will be two rollback threads in parallel. I am not sure if that will work well on MariaDB. It doesn't work on MySQL. We had another patch to initialize more than one rollback threads, thus the big transaction's rollback will not block the rollback of small transactions. There are some statistic variables are global. That need to be fix. 

Second, when server shutdown, it will wait until all active transactions are removed from trx_list. However, the server may exit before ddl_recovery_done, thus there is no chance to start a rollback thread to clear active transactions.

Reply:
It seems it is safe to create multiple rollback thread on MariaDB. I will try to follow your suggestions.

Reply:
As we discussed, This patch will reuse the original recovery rollback thread.

Reply:
renamed the function to innobase_tc_log_recovery_done(). and rename recover_complet() handler interface to signal_tc_log_recovery_done(). Because the function has to be called even binlog is off. background rollback thread has to be started while binlog is off.
HAVE_REPLICATION is for slave code, so it cannot be used here

Reply:
renamed the function to innobase_tc_log_recovery_done(). and rename recover_complet() handler interface to signal_tc_log_recovery_done(). Because the function has to be called even binlog is off. background rollback thread has to be started while binlog is off.
HAVE_REPLICATION is for slave code, so it cannot be used here